pkgs/build-support/node/fetch-yarn-deps/default.nix

```diff
@@ -132,22 +135,49 @@ in
             dontInstall = true;
 
             nativeBuildInputs = [
-              prefetch-yarn-deps
+              (
+                {
+                  "1" = prefetch-yarn-deps;
+                  "3" = yarn-berry_3;
+                  "4" = yarn-berry_4;
+                }
+                .${builtins.toString yarnVersion}
+              )
               cacert
             ];
             GIT_SSL_CAINFO = "${cacert}/etc/ssl/certs/ca-bundle.crt";
             NODE_EXTRA_CA_CERTS = "${cacert}/etc/ssl/certs/ca-bundle.crt";
 
-            buildPhase = ''
-              runHook preBuild
+            configurePhase = lib.optionalString (yarnVersion > 1) ''
+              runHook preConfigure
 
-              yarnLock=''${yarnLock:=$PWD/yarn.lock}
-              mkdir -p $out
-              (cd $out; prefetch-yarn-deps --verbose --builder $yarnLock)
+              export HOME="$NIX_BUILD_TOP"
+              export YARN_ENABLE_TELEMETRY=0
 
-              runHook postBuild
+              yarn config set enableGlobalCache false
+              yarn config set cacheFolder $out
+              yarn config set supportedArchitectures --json "$(cat ${./yarn-berry-supported-archs.json})"
+
+              runHook postConfigure
             '';
 
+            buildPhase =
+              ''
+                runHook preBuild
+
+                mkdir -p $out
+              ''
+              + lib.optionalString (yarnVersion > 1) ''
+                yarn install --immutable --mode skip-build
```

Comment:
This will break the FODs almost instantly if Yarn changes anything in how they pull deps, and also is probably not reproducible between platforms when platform-specific dependencies are used.

Reply:
Valid point. This is partly why there are two yarn versions.

Also this could be said for all external package managers, not just yarn-berry.

I'm uncertain about your point about platform specific dependencies. I'll have to look into it.

Reply:
Platform specific dependencies should be made reproducible by explicitly setting `supportedArchitectures`. Your point about Yarn breaking how they fetch deps may be valid depending on the upstream stability guarantees. Explicitly picking the version of yarn somewhat mitigates that issue, but it's no guarantee yes. This is the same path taken in `pnpm.fetchDeps` #290715